### PR TITLE
Better stream table support over Barrage and jsapi

### LINKED ISSUE #1492: stream table semantics via Barrage
If a stream table has a fast update cycle, and a remote subscriber has a slow update cycle; it would be nice if the update to the remote subscriber was the aggregate of the underlying stream changes.

### LINKED ISSUE #2377: Preserve stream semantics over barrage
It would be nice if stream table semantics were preserved over barrage.

This would allow nice data-mesh applications where workers exporting stream tables could be purpose-built as ingress/egress points.

Discussed with @nbauernfeind and @rcaudy, the implementation would likely be easy.

### PR DESCRIPTION
**Pull request by nbauernfeind**
Fixes #2377. Fixes #1492.

It also adds some stream table specific changes to jsTable that prevent flickering and improve overall GUI experience for stream tables.

### CODE REVIEW

File: server/src/main/java/io/deephaven/server/barrage/BarrageMessageProducer.java [hunk 1]
```diff
@@ -1315,7 +1334,20 @@ private void updateSubscriptionsSnapshotAndPropagate() {
...
+                if (!isStreamTable) {
+                    snapshot = getSnapshot(growingSubscriptions, snapshotColumns, snapshotRowSet,
+                            reverseSnapshotRowSet);
+                } else {
+                    // acquire an empty snapshot to properly align column subscription changes to a UGP step
+                    snapshot = getSnapshot(growingSubscriptions, snapshotColumns, RowSetFactory.empty(),
+                            RowSetFactory.empty());
+
+                    // in the event that the stream table was not empty; pretend it was
+                    if (!snapshot.rowsAdded.isEmpty()) {
```
Comment (lbooker42) [inline server/src/main/java/io/deephaven/server/barrage/BarrageMessageProducer.java:1346]: Is this possible after passing `RowSetFactory.empty()` into `getSnapshot()`?

**Review by lbooker42 — COMMENTED** [review]
These changes look good and are easy to follow, would like to see unit tests exercising this code in `BarrageMessageProducer` and the associated `BarrageTable`s . The `web` changes should be reviewed by @niloc132.

File: web/client-api/src/main/java/io/deephaven/web/client/api/JsTable.java [hunk 2]
```diff
@@ -490,6 +499,11 @@ public void setInternalViewport(double firstRow, double lastRow, Column[] column
         if (firstRow < 0) {
             throw new IllegalArgumentException(firstRow + " < " + 0);
         }
+        final ActiveTableBinding sub = state().getActiveBinding(this);
+        if (isStreamTable() && sub != null && sub.getSubscription() != null) {
```
Comment (niloc132) [inline web/client-api/src/main/java/io/deephaven/web/client/api/JsTable.java:503]: this feels very racy to me, having gotten bitten by this in the past. i think instead you want to test for `subscriptions.get(getHandle())`, and if non-null, check the status of that subscription...

...that said, isn't this still a ui issue, where the UI is calling setViewport unnecessarily?

--

i'd also try to push this into the TableViewportSubscription, which should manage the "is this right one?" question, and just let you say "in this case, never forward the viewport change". If i'm reading this right, you are going to end up passing through there anyway.

File: web/client-api/src/main/java/io/deephaven/web/client/api/JsTable.java [hunk 3]
```diff
@@ -212,6 +217,10 @@ public boolean hasInputTable() {
         return hasInputTable;
     }
 
+    public boolean isStreamTable() {
```
Comment (niloc132) [inline web/client-api/src/main/java/io/deephaven/web/client/api/JsTable.java:220]: if js should know about this, mark it as @JsProperty

File: server/src/main/java/io/deephaven/server/barrage/BarrageMessageProducer.java [hunk 4]
```diff
@@ -2133,17 +2219,23 @@ BarrageMessage getSnapshot(
             final BitSet columnsToSnapshot,
             final RowSet positionsToSnapshot,
             final RowSet reversePositionsToSnapshot) {
-        if (onGetSnapshot != null) {
+        if (onGetSnapshot != null && onGetSnapshotIsPreSnap) {
             onGetSnapshot.run();
         }
 
         // TODO: Use *this* as snapshot tick source for fail fast.
         // TODO: Let notification-indifferent use cases skip notification test
```
Comment (niloc132) [inline server/src/main/java/io/deephaven/server/barrage/BarrageMessageProducer.java:2227]: nit, these are missing issues

File: extensions/barrage/src/main/java/io/deephaven/extensions/barrage/util/BarrageUtil.java [hunk 5]
```diff
@@ -388,12 +392,65 @@ private static ConvertedArrowSchema convertArrowSchema(
...
+        final HashMap<String, String> attributeTypeMap = new HashMap<>();
         tableMetadataVisitor.accept((key, value) -> {
-            final String prefix = ATTR_DH_PREFIX + ATTR_ATTR_TAG + ".";
-            if (!key.startsWith(prefix)) {
+            final String isAttributePrefix = ATTR_DH_PREFIX + ATTR_ATTR_TAG + ".";
+            final String isAttributeTypePrefix = ATTR_DH_PREFIX + ATTR_ATTR_TYPE_TAG + ".";
+            if (key.startsWith(isAttributePrefix)) {
+                result.attributes.put(key.substring(isAttributePrefix.length()), value);
+            } else if (key.startsWith(isAttributeTypePrefix)) {
+                attributeTypeMap.put(key.substring(isAttributeTypePrefix.length()), value);
```
Comment (niloc132) [inline extensions/barrage/src/main/java/io/deephaven/extensions/barrage/util/BarrageUtil.java:403]: looks unrelated, would you consider breaking this out and/or also including this in the js impl, so that it can have support for knowing what the types of attrs should be?

File: web/client-api/src/main/java/io/deephaven/web/client/api/JsTable.java [hunk 6]
```diff
@@ -1220,6 +1234,18 @@ public void handleDelta(ClientTableState current, DeltaUpdates updates) {
                 nonViewportSub.handleDelta(updates);
                 return;
             }
+            if (isStreamTable) {
+                // stream tables remove all rows from the previous step, if there are no adds this step then defer
+                // removal until new data arrives -- this makes stream tables GUI friendly
+                if (updates.getAdded().isEmpty()) {
+                    deferredDeltaRemoves = updates.getRemoved();
```
[the remark this answers is not in the document]
Comment (nbauernfeind) [reply web/client-api/src/main/java/io/deephaven/web/client/api/JsTable.java:1241]: added an assert

File: server/src/main/java/io/deephaven/server/barrage/BarrageMessageProducer.java (hunk 4 above)
Comment (nbauernfeind) [reply server/src/main/java/io/deephaven/server/barrage/BarrageMessageProducer.java:2227]: Looks like Ryan added them; I'll keep up with him to figure something out.

File: web/client-api/src/main/java/io/deephaven/web/client/api/barrage/def/TableAttributesDefinition.java [hunk 7]
```diff
@@ -41,6 +45,10 @@ public boolean isInputTable() {
         return remainingAttributeKeys.contains(INPUT_TABLE_ATTRIBUTE);
     }
 
+    public boolean isStreamTable() {
+        return map.getOrDefault(STREAM_TABLE_ATTRIBUTE, "").equals("true");
```
Comment (niloc132) [inline web/client-api/src/main/java/io/deephaven/web/client/api/barrage/def/TableAttributesDefinition.java:49]: consider `"true".equals(map.get(STREAM_TABLE_ATTRIBUTE)` so you don't have to care about nulls

File: web/client-api/src/main/java/io/deephaven/web/client/api/JsTable.java [hunk 8]
```diff
@@ -1220,6 +1230,20 @@ public void handleDelta(ClientTableState current, DeltaUpdates updates) {
                 nonViewportSub.handleDelta(updates);
                 return;
             }
+            if (isStreamTable) {
```
Comment (niloc132) [inline web/client-api/src/main/java/io/deephaven/web/client/api/JsTable.java:1233]: This whole block should be able to move upstream too, so that if there is more than one Table that is attached to a single ticket, they all get their removes deferred at once, instead of each tracking its own deferrals?

ClientTableState.handleDelta is before it fans out to each downstream sub (admittedly it is weird to hold state about a subscription there), or WorkerConnection.incrementalUpdates is one step upstream of that. I could be pretty easily convinced that neither is appropriate.

Upstream another step is the anonymous class that handles FlightData messages arriving - this _does_ track state of messages, presently to be able to see how many more messages are needed before the current delta is done. This use case smells a bit like that?

**Review by niloc132 — APPROVED** [review]
Marked approved, but with one "maybe change this, it seems potentially better to me" and one "this could be simpler if you wanted", but both optional.
